File: solutions/src/MinStack.java
import java.util.Stack;

public class MinStack {

    private Stack<Integer> minStack;
    private Stack<Integer> regularStack;
    private Integer min;
    public MinStack() {
        minStack = new Stack<>();
        regularStack = new Stack<>();
        min = null;
    }

    public void push(int x) {
        regularStack.push(x);
        if (min == null || min >= x) {
            minStack.push(x);
            min = x;
        }
    }
    public int getMin() {
        return min;
    }
    public int top() {
        return regularStack.peek();
    }
    public int pop() {
        int toRemove = regularStack.pop();
        if (min == toRemove) {
            if (regularStack.size() == 0) {
                minStack.pop();
                min = null;
            }
            else {
                minStack.pop();
                min = minStack.peek();
            }
        }
        return toRemove;
    }
}

File: solutions/src/MinStack.kt
/**
 * https://leetcode.com/problems/min-stack/submissions/
 */
class MinStack {

    val minStack = mutableListOf<Int>()
    val orderStack = mutableListOf<Int>()
    fun push(`val`: Int) {
        orderStack.add(`val`)
        if (minStack.isEmpty()) {
            minStack.add(`val`)
        }
        else if (minStack.first() > `val`) {
            minStack.add(`val`)
        }
        else {
            minStack.add(minStack.last())
        }
    }

    fun pop() {
        if (orderStack.isNotEmpty()) {
            minStack.removeAt(minStack.lastIndex)
            orderStack.removeAt(orderStack.lastIndex)
        }
    }

    fun top(): Int {
        return orderStack.last()
    }

    fun getMin(): Int {
        return minStack.last()
    }
}